sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/loss.py

```diff
@@ -37,42 +41,26 @@ class RegressorLossReduction(BERTMLMLossWithReduction):
 
     def forward(
         self, batch: Dict[str, Tensor], forward_out: Dict[str, Tensor]
-    ) -> Tuple[Tensor, PerTokenLossDict | SameSizeLossDict]:
-        """Calculates the loss within a micro-batch. A micro-batch is a batch of data on a single GPU.
+    ) -> Tuple[torch.Tensor, torch.Tensor, Dict[str, torch.Tensor]]:
+        """Calculates the sum of squared errors within a micro-batch. A micro-batch is a batch of data on a single GPU. The averaging of the loss is done in super().reduce().
 
         Args:
             batch: A batch of data that gets passed to the original forward inside LitAutoEncoder.
             forward_out: the output of the forward method inside classification head.
-
-        Returns:
-            A tuple containing [<loss_tensor>, ReductionT] where the loss tensor will be used for
-                backpropagation and the ReductionT will be passed to the reduce method
-                (which currently only works for logging.).
         """
         regression_output = forward_out["regression_output"]
         targets = batch["labels"].to(dtype=regression_output.dtype)  # [b, 1]
 
+        num_valid_tokens = torch.tensor(targets.numel(), dtype=torch.int, device=targets.device)
+
         cp_size = parallel_state.get_context_parallel_world_size()
         if cp_size == 1:
-            loss = torch.nn.functional.mse_loss(regression_output, targets)
+            loss_sum = ((regression_output - targets) ** 2).sum()  # [b, 1]
```

Comment:
Are we sure by this point that the final linear layer is not column parallel? I assume probably not since the final dim is something like `[b, s, 1]` prior to this sum reduction, right? Also see my suggestion below, as is you are also reducing over the batch with the default `.sum()`
```suggestion
            loss_sum = ((regression_output - targets) ** 2).sum(dim=-1)  # [b, 1]
```

Reply:
My question about column parallel is related to whether or not this loss is valid for tensor-parallel models? The cross entropy function you use in the classification loss _is_ TP safe, so that should work, but the F.cross_entropy that used to be there was not TP safe. This may still be TP unsafe, unless we know for sure that the different TP ranks are all getting the same linear layer result. We may need to do something like te.Linear and set `weight.sequence_parallel=True` or something so it gets all reduced across TP ranks... There's hackery like this inside of the Evo2 custom layers if we want to verify.

Reply:
https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py#L507-L516 setting this magic parameter to `True` tells Megatron that the parameter _should not_ be reduced, so... maybe by default things _will_ be reduced?

Reply:
actually i just debugged on the minimal example and it looks correct how it is right now, when i use TP=2 the loss is the same as TP=1 across multiple steps, and when i use DP=2 the loss is 1/2 as DP=1. This also means loss for logging stays the same for all

Reply:
Sounds good, let's keep our eyes on the slack thread and if it turns out it is subtly different, or keeps things from drifting better, then we can go back to it. I've seen things where grads are _slightly_ different between ranks if they are not explicitly mean reduced, and that may not be happening now. It may just work for now because they start the same and run on the same data.

Reply:
https://nvidia.slack.com/archives/C0434FDLPQV/p1751059626701659 or https://nvidia.slack.com/archives/C01JPHFK7DH/p1751054240388109